### ISSUE TITLE
runTest Aff resolves before tests finish

### ISSUE THREAD

**Issue by lpil** [bodil/purescript-test-unit#39]
Hello!

I wish to run some clear-up code after my test suite, but I'm finding that the runTest function returns before the tests have finished running. Here is a reproduction:

 ```purescript
main =
  runTest suites
  log "##################### This should be after"
```

I would expect this output:

```
* Running tests...
→ Suite: Web.app
  → Running: test 1✓ Passed: test 1
  → Running: test 2✓ Passed: test 2
  → Running: test 3✓ Passed: test 3
  → Running: test 4✓ Passed: test 4

All 4 tests passed! 🎉
##################### This should be after
```

Instead I am getting this output:

```
* Running tests...
→ Suite: Web.app
  → Running: test 1✓ Passed: test 1
  → Running: test 2✓ Passed: test 2
  → Running: test 3✓ Passed: test 3
  → Running: test 4##################### This should be after
✓ Passed: test 4

All 4 tests passed! 🎉
```

Am I using `runTest` incorrectly? How might I get the expected behaviour detailed above?

Thanks,
Louis

**Comment by mattaudesse** [comment]
Hey @lpil - I know this is an old question but I'll try to answer in case you didn't figure it out:

Since `runTest suites` is executed asynchronously and discards its resultant `Aff` (you get back an `Eff e Unit` instead), I think the behavior you're seeing makes sense. In other words, `runTest` is performing as it should based on its type signature, but you can't use it to do what you're trying to achieve above.

Luckily there's [`runTestWith`](https://pursuit.purescript.org/packages/purescript-test-unit/13.0.0/docs/Test.Unit.Main#v:runTestWith) which *does* return its `Aff`, so that should allow you to explicitly sequence your
```purescript
log "##################### This should be after"
```
line to run *afterward* by changing the body of `main` to also run in `Aff`, or by using one of the helpers from the [`Control.Monad.Aff` library](https://pursuit.purescript.org/packages/purescript-aff/4.0.2/docs/Control.Monad.Aff) (perhaps `finally`).

Here's an example from the [`Aff` docs](https://github.com/slamdata/purescript-aff#example)
```purescript
main = launchAff do
  response <- Ajax.get "http://foo.bar"
  log response.body
```

In your case, I believe (but haven't tested) that this would work:
```purescript
import Test.Unit.Output.Fancy as Fancy

main = launchAff do
  _ <- runTestWith Fancy.runTest suites
  log "##################### This should be after"
```

**tl;dr sequencing behavior boils down to whether you're running inside `Eff` or `Aff`.**

**NB** Also note that test failures trigger `exit 1` in the running process, so your line of hashes wouldn't be printed unless all your tests were "in the green".

**Comment by lpil** [comment]
Wonderful, a clear explanation. Thank you for the help :)
